cookbooks/aws-parallelcluster-slurm/recipes/update_shared_storages.rb

```diff
@@ -0,0 +1,181 @@
+# License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+load_shared_storages_mapping
+
+ruby_block "get volumes to mount and unmount" do
+  block do
+    shared_storages_mapping = node['cluster']['shared_storages_mapping']
+    update_shared_storages_mapping = node['cluster']['update_shared_storages_mapping']
+    # get ebs to unmount
+    unmount_shared_dir_array = []
+    unmount_vol_array = []
+    unless shared_storages_mapping["ebs"].nil?
+      shared_storages_mapping["ebs"].each do |volume|
+        next unless update_shared_storages_mapping["ebs"].nil? || !update_shared_storages_mapping["ebs"].include?(volume)
+        unmount_shared_dir_array.push(volume["mount_dir"])
+        unmount_vol_array.push(volume["volume_id"])
+      end
+    end
+    node.default['cluster']['unmount_shared_dir_array'] = unmount_shared_dir_array
+    node.default['cluster']['unmount_vol_array'] = unmount_vol_array
+
+    # get ebs to mount
+    mount_shared_dir_array = []
+    mount_vol_array = []
+    unless update_shared_storages_mapping["ebs"].nil?
+      update_shared_storages_mapping["ebs"].each do |volume|
+        next unless shared_storages_mapping["ebs"].nil? || !shared_storages_mapping["ebs"].include?(volume)
+        mount_shared_dir_array.push(volume["mount_dir"])
+        mount_vol_array.push(volume["volume_id"])
+      end
+    end
+    node.default['cluster']['mount_shared_dir_array'] = mount_shared_dir_array
+    node.default['cluster']['mount_vol_array'] = mount_vol_array
+
+    # get efs to unmount
+    unmount_efs_fs_id_array = []
+    unmount_efs_shared_dir_array = []
+    unless shared_storages_mapping["efs"].nil?
+      shared_storages_mapping["efs"].each do |volume|
+        next unless update_shared_storages_mapping["efs"].nil? || !update_shared_storages_mapping["efs"].include?(volume)
+        unmount_efs_shared_dir_array.push(volume["mount_dir"])
+        unmount_efs_fs_id_array.push(volume["volume_id"])
+      end
+    end
+    node.default['cluster']['unmount_efs_shared_dir_array'] = unmount_efs_shared_dir_array
+    node.default['cluster']['unmount_efs_fs_id_array'] = unmount_efs_fs_id_array
+
+    # get efs to mount
+    mount_efs_fs_id_array = []
+    mount_efs_shared_dir_array = []
+    unless update_shared_storages_mapping["efs"].nil?
+      update_shared_storages_mapping["efs"].each do |volume|
+        next unless shared_storages_mapping["efs"].nil? || !shared_storages_mapping["efs"].include?(volume)
+        mount_efs_shared_dir_array.push(volume["mount_dir"])
+        mount_efs_fs_id_array.push(volume["volume_id"])
```

Comment:
The logic to detect added and removed entries is duplicated for EBS and EFS. What about reducing this code duplication?

Reply:
Created different functions to handle EBS, EFS, FSX mount and unmount to reduce code duplication